nilearn/input_data/base_masker.py

```diff
@@ -5,75 +5,95 @@
 # License: simplified BSD
 
 import warnings
+import abc
 
 import numpy as np
-import itertools
 
 from sklearn.base import BaseEstimator, TransformerMixin
-from sklearn.externals.joblib import Memory, Parallel, delayed
+from sklearn.externals.joblib import Memory
 
 from .. import masking
 from .. import image
 from .. import signal
 from .. import _utils
 from .._utils.cache_mixin import CacheMixin, cache
-from .._utils.class_inspect import enclosing_scope_name, get_params
-from .._utils.compat import _basestring, izip
-from nilearn._utils.niimg_conversions import (
-    _iter_check_niimg, _check_same_fov)
-
-
-def filter_and_mask(imgs, mask_img_,
-                    parameters,
-                    memory_level=0,
-                    memory=Memory(cachedir=None),
-                    verbose=0,
-                    confounds=None,
-                    copy=True,
-                    sample_mask=None):
-    # If we have a string (filename), we won't need to copy, as
-    # there will be no side effect
+from .._utils.class_inspect import enclosing_scope_name
+from .._utils.compat import _basestring
 
-    if isinstance(imgs, _basestring):
-        copy = False
 
-    if verbose > 0:
-        class_name = enclosing_scope_name(stack_level=2)
+def filter_and_extract(imgs, extraction_function,
+                       parameters,
+                       memory_level=0, memory=Memory(cachedir=None),
+                       verbose=0,
+                       confounds=None,
+                       copy=True):
+    """Extract representative time series using given function.
 
-    mask_img_ = _utils.check_niimg_3d(mask_img_)
+    Parameters
+    ----------
+    imgs: 3D/4D Niimg-like object
+        Images to be masked. Can be 3-dimensional or 4-dimensional.
 
-    imgs = _utils.check_niimg(imgs, atleast_4d=True, ensure_ndim=4)
-    if sample_mask is not None:
-        imgs = image.index_img(imgs, sample_mask)
+    extraction_function: function
+        Function used to extract the time series from 4D data. This function
+        should take images as argument and returns a 2D array with masked
+        signals. If any other parameter is needed, a functor or a partial
```

Comment:
We can't do that with a functor since the operation has to remain stateless because of the caching system.